### PR TITLE
[SourceLink] Create output window for logging, and tell the user what is going on

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/57352
Part of #55834
Includes #57978 so just review [`33072b9` (#57996)](https://github.com/dotnet/roslyn/pull/57996/commits/33072b9b176d16a6913152a036529b7f478beb44)

The idea is to help library authors work out why their embedded or source link sources might not be working, we can provide this output window. It will also be a good source for us to ask for information from, for feedback issues.

Screenshots:
image

image

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            OutputPane.Clear();
+        }
+
+        public void Log(string message)
+        {
+            OutputPane.WriteLine(message);
+        }
+
+        private class OutputPane
```
Comment (ryzngard) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:36]: Not sure I follow why this is an embedded class... it's only statically used. Could just remove that layer and move all members up

**Review by ryzngard — CHANGES_REQUESTED** [review]

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs (hunk 1 above)
Comment (davidwengier) [reply src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:36]: Yeah, fair call. This is just a copy and paste, thank you for keeping me honest. Will clean it up :)

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 2]
```diff
@@ -2948,4 +2948,61 @@ Zero-width positive lookbehind assertions are typically used at the beginning of
   <data name="Silent" xml:space="preserve">
     <value>Silent</value>
   </data>
+  <data name="Count_not_find_PDB_on_disk_or_embedded" xml:space="preserve">
+    <value>Count not find portable PDB on disk, or embedded.</value>
```
[the remark this answers is not in the document]
Comment (davidwengier) [reply src/Features/Core/Portable/FeaturesResources.resx:2952]: Could -> Count is one of my most common typos. My fingers have a mind of their own.

**Comment by davidwengier** [comment]
Updated. Sorry I had to rebase because of changes to the other PR

**Comment by davidwengier** [comment]
> Updated. Sorry I had to rebase, and therefore force push, because of changes to the other PR

Just realised I'll probably be rebasing again when that PR merges 🤦‍♂️

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public PdbSourceDocumentOutputWindowLogger(SVsServiceProvider serviceProvider, IThreadingContext threadingContext)
+        {
+            _serviceProvider = serviceProvider;
+            _threadingContext = threadingContext;
+        }
+
+        public void Clear()
+        {
+            _threadingContext.JoinableTaskFactory.Run(async () =>
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:35]: Should these methods be async so we're not potentially blocking up a background thread?

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                var outputWindow = (IVsOutputWindow)_serviceProvider.GetService(typeof(SVsOutputWindow));
+
+                // this should bring outout window to the front
+                _outputPane = CreateOutputPane(outputWindow);
+            }
+
+            return _outputPane;
+        }
+
+        private IVsOutputWindowPane? CreateOutputPane(IVsOutputWindow outputWindow)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:76]: TryCreate...?

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            _threadingContext.ThrowIfNotOnUIThread();
+
+            // Try to get the workspace pane if it has already been registered
+            var workspacePaneGuid = s_outputPaneGuid;
+
+            // If the pane has already been created, CreatePane returns it
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:83]: Comment doesn't make sense -- GetPane is returning it...?

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                GetPane()?.Clear();
+            });
+        }
+
+        public void Log(string value)
+        {
+            _threadingContext.JoinableTaskFactory.Run(async () =>
+            {
+                // Despite being named "ThreadSafe" unfortunately OutputStringThreadSafe does not play well with JTF
+                // which the debugger needs for SourceLink
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:48]: The docs for the "ThreadSafe" method imply it's for pumping reasons, and there's some advice about casting to a different interface instead. Should we be doing that?

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var pane = GetPane();
+                if (pane == null)
+                {
+                    return;
+                }
+
+                pane.OutputStringThreadSafe(value + Environment.NewLine);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:57]: pane?.OutputStringThreadSafe...?

File: src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/PdbSourceDocument/PdbSourceDocumentOutputWindowLogger.cs:35]: I also worry about deadlock risk here if we end up with a UI thread blocking on the call to GetGeneratedFileAsync, and then this trying to hop back. Given this is only logging, should this just be queuing work to the UI thread and we keep the actual updating of the pane asynchronous? It'd remove the deadlock risk, and also would probably help performance so you're not doing so many UI transitions during the main code path.

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs [hunk 8]
```diff
@@ -54,22 +54,29 @@ internal sealed class PdbSourceDocumentMetadataAsSourceFileProvider : IMetadataA
             if (signaturesOnly)
                 return null;
 
+            var assemblyName = symbol.ContainingAssembly.Identity.Name;
+
+            _logger?.Clear();
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:59]: It's implicit we're never doing two calls to GetGeneratedFIleAsync at once?

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs [hunk 9]
```diff
@@ -54,22 +54,29 @@ internal sealed class PdbSourceDocumentMetadataAsSourceFileProvider : IMetadataA
...
+            _logger?.Log(FeaturesResources.Navigating_to_symbol_0_from_1, symbol, assemblyName);
+
             // If this is a reference assembly then we won't have the right information available, so bail out
             // TODO: find the implementation assembly for the reference assembly, and keep going: https://github.com/dotnet/roslyn/issues/55834
             var isReferenceAssembly = symbol.ContainingAssembly.GetAttributes().Any(attribute => attribute.AttributeClass?.Name == nameof(ReferenceAssemblyAttribute)
                 && attribute.AttributeClass.ToNameDisplayString() == typeof(ReferenceAssemblyAttribute).FullName);
             if (isReferenceAssembly)
+            {
+                _logger?.Log("");
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:68]: Why?

File: src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs [hunk 10]
```diff
@@ -54,22 +54,29 @@ internal sealed class PdbSourceDocumentMetadataAsSourceFileProvider : IMetadataA
...
                 return null;
+            }
 
             var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
 
             if (compilation.GetMetadataReference(symbol.ContainingAssembly) is not PortableExecutableReference { FilePath: not null and var dllPath })
             {
+                _logger?.Log(FeaturesResources.Source_is_a_reference_assembly);
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentMetadataAsSourceFileProvider.cs:76]: Did you mean to put this earlier? The message here should probably be different.

**Comment by davidwengier** [comment]
Closing this. Going to combine with the document tab changes, and add some telemetry, in a follow up since this missed Preview 2 anyway.

**Pull request closed without merging**